README
NAME
    App::assh - A wrapper around autossh.

VERSION
    version 1.1.1

SYNOPSIS
    A wrapper around autossh.

MOTIVATION
    `autossh` is a nifty little ssh-keepalive-connection-holder.

    Passing in the ports for the keepalive can be clumsy though: `assh`
    helps you to avoid that.

USAGE
         assh
         assh HOSTNAME

REQUIREMENTS
    First, you will need a file `~/.ssh/config`. It looks something like
    this:

         Host foo
         HostName bar.example.com
         User baz

    With this, you can already leverage standard `ssh` connections:

         ssh foo

    ... instead of

         ssh baz@bar.example.com

    Next, generate a file `~/.autossh_rc` with the following format:

         foo 12345

    ... with the first entry on the line representing your `Host` in
    `~/.ssh/config` and the second item on the line being the port over
    which to keep the autossh connection alive.

    Now you can permanently connect using:

         assh foo

    ... with the connection kept alive across network switches and computer
    shutdowns.

ATTRIBUTES
    *   hosts: HashRef holding the values HOSTNAME => AUTOSSH_PORT

    *   ports: HashRef holding the values HOSTNAME => "USER => USERNAME,
        HOST => HOSTNAME"

    *   ssh_config_file: The path to the ssh config file. Default:
        `~/.ssh/config`

    *   ports_config_file: The path to the ports config (this is what I have
        chosen): `~/.autossh_rc`

SEE ALSO
    *   autossh: <http://www.harding.motd.ca/autossh/>

AUTHOR
    DBR <dbr@cpan.org>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2013 by DBR.

    This is free software, licensed under:

      DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE, Version 2, December 2004



Changes
1.1.1     2013-09-04 18:15:34 Europe/Berlin

    * Version 1.1.0 was rejected by cpan

1.1.0     2013-09-04 10:07:00 Europe/Berlin

    * Fix documentation issue.

1.0.0

    * Addded more documentation

Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "A wrapper around autossh.",
  "dist_author" => [
    "DBR <dbr\@cpan.org>"
  ],
  "dist_name" => "App-assh",
  "dist_version" => "1.1.1",
  "license" => "unrestricted",
  "module_name" => "App::assh",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "FindBin" => 0,
    "Moo" => 0,
    "MooX::Options" => 0,
    "MooX::Types::MooseLike::Base" => 0,
    "lib" => 0,
    "methods" => 0,
    "perl" => "v5.10.0",
    "strict" => 0,
    "true" => 0,
    "warnings" => 0
  },
  "script_files" => [
    "bin/assh",
    "bin/assh"
  ],
  "test_requires" => {
    "Carp" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec::Functions" => 0,
    "List::Util" => 0,
    "Scalar::Util" => 0,
    "Test::CheckDeps" => "0.006",
    "Test::More" => "0.94",
    "Test::UseAllModules" => 0
  }
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/App/assh.pm
package App::assh;
BEGIN {
  $App::assh::AUTHORITY = 'cpan:DBR';
}
{
  $App::assh::VERSION = '1.1.1';
}

#  PODNAME: App::assh
# ABSTRACT: A wrapper around autossh.

use Moo;
use true;
use 5.010;
use strict;
use warnings;
use methods-invoker;
use MooX::Options skip_options => [qw<os cores>];
use MooX::Types::MooseLike::Base qw(:all);

has hosts => (
    is => 'lazy',
    isa => HashRef,
);

has ports => (
    is => 'lazy',
    isa => HashRef,
);

has ssh_config_file => (
    is => 'ro',
    isa => Str,
    default => sub {
        "$ENV{HOME}/.ssh/config"
    },
);

has ports_config_file => (
    is => 'ro',
    isa => Str,
    default => sub {
        "$ENV{HOME}/.autossh_rc"
    },
);

method _build_hosts {
    $_ = do { local(@ARGV, $/) = $->ssh_config_file; <>; };
    s/\s+/ /g;

    my $ret = {};
    while (m<Host\s(.+?)\sHostName\s(.+?)\sUser\s(.+?)\s>xg) {
        $ret->{$1} = { NAME => $2, USER => $3 }
    }

    return $ret;
}

method _build_ports {
    open my $portsfile, "<", $->ports_config_file or die $!;
    my $h = {};
    while(<$portsfile>) {
        chomp;
        my ($host, $port) = split;
        $h->{$host} = $port;
    }
    return $h
}

method run {
    my $host = shift;

    not defined $host and do {
        say for keys %{$->hosts};
    };

    defined $->hosts->{$host} and do {
        $->autossh_exec($host);
    };
}

method autossh_exec {
    my $host = shift;
    exec 'AUTOPOLL=5 autossh -M ' . $->ports->{$host} . ' ' . $->hosts->{$host}{USER} . '@' . $->hosts->{$host}{NAME}
}

no Moo;

__END__

=pod

=encoding utf-8

=head1 NAME

App::assh - A wrapper around autossh.

=head1 VERSION

version 1.1.1

=head1 SYNOPSIS

A wrapper around autossh.

=for Pod::Coverage ports_config_file  ssh_config_file

=head1 MOTIVATION

`autossh` is a nifty little ssh-keepalive-connection-holder.

Passing in the ports for the keepalive can be clumsy though: `assh` helps you to avoid that.

=head1 USAGE

     assh
     assh HOSTNAME

=head1 REQUIREMENTS

First, you will need a file `~E<sol>.sshE<sol>config`. It looks something like this:

     Host foo
     HostName bar.example.com
     User baz

With this, you can already leverage standard `ssh` connections:

     ssh foo

... instead of

     ssh baz@bar.example.com

Next, generate a file `~E<sol>.autossh_rc` with the following format:

     foo 12345

... with the first entry on the line representing your `Host` in `~E<sol>.sshE<sol>config` and the second item on the line being the port over which to keep the autossh connection alive.

Now you can permanently connect using:

     assh foo

... with the connection kept alive across network switches and computer shutdowns.

=head1 ATTRIBUTES

=over

=item *

hosts: HashRef holding the values HOSTNAME =E<gt> AUTOSSH_PORT

=back

=over

=item *

ports: HashRef holding the values HOSTNAME =E<gt> C<<< USER => USERNAME, HOST => HOSTNAME >>>

=back

=over

=item *

ssh_config_file: The path to the ssh config file. Default: `~E<sol>.sshE<sol>config`

=back

=over

=item *

ports_config_file: The path to the ports config (this is what I have chosen): `~E<sol>.autossh_rc`

=back

=head1 SEE ALSO

=over

=item *

autossh: E<lt>http:E<sol>E<sol>www.harding.motd.caE<sol>autosshE<sol>E<gt>

=back

=head1 AUTHOR

DBR <dbr@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2013 by DBR.

This is free software, licensed under:

  DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE, Version 2, December 2004

=cut


Makefile.PL

use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "A wrapper around autossh.",
  "AUTHOR" => "DBR <dbr\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "App-assh",
  "EXE_FILES" => [
    "bin/assh",
    "bin/assh"
  ],
  "LICENSE" => "unrestricted",
  "NAME" => "App::assh",
  "PREREQ_PM" => {
    "FindBin" => 0,
    "Moo" => 0,
    "MooX::Options" => 0,
    "MooX::Types::MooseLike::Base" => 0,
    "lib" => 0,
    "methods" => 0,
    "strict" => 0,
    "true" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Carp" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec::Functions" => 0,
    "List::Util" => 0,
    "Scalar::Util" => 0,
    "Test::CheckDeps" => "0.006",
    "Test::More" => "0.94",
    "Test::UseAllModules" => 0
  },
  "VERSION" => "1.1.1",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



